### PR TITLE
Fix double handling in `collect_tokens`

### PR DESCRIPTION
**Pull request by nnethercote**
Double handling of AST nodes can occur in `collect_tokens`. This is when an inner call to `collect_tokens` produces an AST node, and then an outer call to `collect_tokens` produces the same AST node. This can happen in a few places, e.g. expression statements where the statement delegates `HasTokens` and `HasAttrs` to the expression. It will also happen more after #124141.

This PR fixes some double handling cases that cause problems, including #129166.

r? @petrochenkov

### CODE REVIEW

**Comment by nnethercote** [comment]
The last three commits are tricky and hard to review. Even after multiple weeks of working with this code, I still find it hard to understand and reason about, and I wish it could be made simpler. But the good news is that these commits are enough to the get final failing test (`tests/ui/proc-macros/macro-rules-derive-cfg.rs`) working in #124141.

**Comment by nnethercote** [comment]
I think this should get a crater run, just because I don't fully trust that the test suite is thoroughly testing `cfg_attr` stuff.

And why not do a perf run, just to check there's no regressions.

@bors try @rust-timer queue

File: tests/ui/conditional-compilation/invalid-node-range-issue-129166.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// This was triggering an assertion failure in `NodeRange::new`.
```
Comment (petrochenkov) [inline tests/ui/conditional-compilation/invalid-node-range-issue-129166.rs:1]: Could you split the assert in `NodeRange::new` into two asserts?
Right now it is not clear which part of the condition actually fails.

**Comment by petrochenkov** [comment]
I'll start crater once https://github.com/rust-lang/rust/pull/129346#discussion_r1725291124 is addressed and CI is green.
@rustbot author

**Comment by nnethercote** [comment]
I have adding a commit that avoids cloning `ret.attrs()` except when necessary. This gets back part of the perf regression. The rest is due to `Parser::clone` being more expensive because of the new `FxHashSet` in `CaptureState`. I tried changing it to a `GrowableBitSet` which seemed like it should have helped but didn't.

Let's see what the new status is:
@bors try @rust-time queue

**Comment by lqd** [comment]
@rust-timer build a70947b8062a88e6f76266f2540de6ea6c2bfba5

**Comment by nnethercote** [comment]
That didn't help much, but I have another idea that I think should work better. I will try it tomorrow. We should hold off on a crater run until then.

**Comment by nnethercote** [comment]
Another attempt to improve perf, this time using a `RangeSet` from the `rangemap` crate.

@bors try @rust-timer queue

**Comment by lqd** [comment]
> this time using a `RangeSet` from the `rangemap` crate.

Drive-by question, is that like `rustc_index`'s `IntervalSet`?

**Comment by nnethercote** [comment]
> Drive-by question, is that like `rustc_index`'s `IntervalSet`?

Oh! It is, thank you for the suggestion. But `IntervalSet` is even better because its API matches that of `HashSet` exactly, at least for the uses in this PR. I have switched the code to use it, let's do another perf run:

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
Ok, `IntervalSet` gives similar results to `RangeMap`, which is great! There is still a very small perf regression but it's less than it was, and the fix is necessary for correctness both now and even more when #124141 gets merged. So I think this is good enough for a crater run.

@bors try

**Comment by nnethercote** [comment]
@craterbot check

**Comment by nnethercote** [comment]
The reason that `IntervalSet` works well is that multi-line `///` doc comments get split into one attribute per line. So it's common to have lots of attributes on an AST node, but those attributes all have consecutive `AttrId`s. Much nicer to have a single range like `50..100` instead of 50 individual entries.

I really wish that multiple `///` doc comment lines instead collapsed into a single attribute. This would make it easier to give good perf inside the compiler. I suspect it's not possible to change this because of proc macros.

**Comment by danielhenrymantilla** [comment]
FWIW, this did regress the following test: https://docs.rs/crate/cfg_eval/0.1.2/source/tests/item.rs

  - https://github.com/danielhenrymantilla/cfg_eval.rs/actions/workflows/future-proof.yml

I don't know why crater did not see it 😔

File: tests/ui/proc-macro/macro-rules-derive-cfg.rs [hunk 2]
```diff
@@ -14,17 +14,15 @@ extern crate test_macros;
 macro_rules! produce_it {
     ($expr:expr) => {
         #[derive(Print)]
-        struct Foo {
-            val: [bool; {
-                let a = #[cfg_attr(not(FALSE), rustc_dummy(first))] $expr;
-                0
-            }]
-        }
```
Comment (danielhenrymantilla) [inline tests/ui/proc-macro/macro-rules-derive-cfg.rs:22]: This case stopped working?

**Comment by Noratrieb** [comment]
even if crater found your crate, it was only a check run

File: tests/ui/proc-macro/macro-rules-derive-cfg.rs (hunk 2 above)
Comment (nnethercote) [reply tests/ui/proc-macro/macro-rules-derive-cfg.rs:22]: Fixed in #132587.

**Comment by lqd** [comment]
> even if crater found your crate, it was only a check run

and since beta runs are test runs, the regression is indeed present in the 1.83 unexpected test failures: https://crater-reports.s3.amazonaws.com/beta-1.83-1/beta-2024-10-19/gh/danielhenrymantilla.cfg_eval.rs/log.txt

**Pull request merged**